### PR TITLE
no_std support for substrate trie

### PR DESCRIPTION
**Pull request by cheme**
This PR allows to use trie with no_std. Solves #1574.

### CODE REVIEW

File: core/client/Cargo.toml [hunk 1]
```diff
@@ -18,7 +18,7 @@ state-machine = { package = "substrate-state-machine", path = "../state-machine"
 keyring = { package = "substrate-keyring", path = "../keyring", optional = true }
 trie = { package = "substrate-trie", path = "../trie", optional = true }
 substrate-telemetry = { path = "../telemetry", optional = true }
-hash-db = { version = "0.12", optional = true }
+hash-db = { version = "0.12", default-features = false }
```
Comment (bkchr) [inline core/client/Cargo.toml]:
```suggestion
hash-db = { version = "0.12.2", default-features = false, optional = true }
```

File: core/sr-io/Cargo.toml [hunk 2]
```diff
@@ -17,19 +17,18 @@ libsecp256k1 = { version = "0.2.1", optional = true }
 tiny-keccak = { version = "1.4.2", optional = true }
 environmental = { version = "1.0.1", optional = true }
 substrate-state-machine = { path = "../state-machine", optional = true }
-trie = { package = "substrate-trie", path = "../trie", optional = true }
+trie = { package = "substrate-trie", path = "../trie", default-features = false }
```
Comment (bkchr) [inline core/sr-io/Cargo.toml]:
```suggestion
trie = { package = "substrate-trie", path = "../trie", default-features = false, optional = true }
```

File: core/test-runtime/src/lib.rs [hunk 3]
```diff
@@ -477,6 +516,36 @@ cfg_if! {
 					panic!("Failing because we are on wasm")
 				}
 
+				fn use_trie() -> u64 {
+					let pairs = [
```
Comment (bkchr) [inline core/test-runtime/src/lib.rs]: Both versions provide the same functionality, please move into one common function and just call this.

And you also don't call these functions somewhere in a test?

File: core/test-runtime/src/lib.rs [hunk 4]
```diff
@@ -477,6 +516,36 @@ cfg_if! {
...
+
+					let mut mdb = PrefixedMemoryDB::default();
+					let mut root = rstd::default::Default::default();
+					let _ = {
+						let v = &pairs;
+						let mut t = TrieDBMut::<Blake2Hasher>::new(&mut mdb, &mut root);
+						for i in 0..v.len() {
+							let key: &[u8]= &v[i].0;
+							let val: &[u8] = &v[i].1;
+							t.insert(key, val).unwrap();
```
Comment (bkchr) [inline core/test-runtime/src/lib.rs]: Would be nice to replace the unwraps with expects.

File: core/trie/Cargo.toml [hunk 5]
```diff
@@ -12,11 +12,12 @@ name = "bench"
 harness = false
 
 [dependencies]
-codec = { package = "parity-codec", version = "3.2" }
+codec = { package = "parity-codec", version = "3.2", default-features = false }
+rstd = { package = "sr-std", path = "../sr-std", default-features = false }
 hash-db = { version = "0.12", default-features = false }
-trie-db = { version = "0.12", optional = true }
+trie-db = { version = "0.12", default-features = false }
```
Comment (bkchr) [inline core/trie/Cargo.toml]:
```suggestion
trie-db = { version = "0.12.2", default-features = false }
```

File: core/trie/Cargo.toml [hunk 6]
```diff
@@ -12,11 +12,12 @@ name = "bench"
 harness = false
 
 [dependencies]
-codec = { package = "parity-codec", version = "3.2" }
+codec = { package = "parity-codec", version = "3.2", default-features = false }
+rstd = { package = "sr-std", path = "../sr-std", default-features = false }
 hash-db = { version = "0.12", default-features = false }
```
Comment (bkchr) [inline core/trie/Cargo.toml]:
```suggestion
hash-db = { version = "0.12.2", default-features = false }
```

File: core/trie/Cargo.toml [hunk 7]
```diff
@@ -12,11 +12,12 @@ name = "bench"
...
-trie-db = { version = "0.12", optional = true }
+trie-db = { version = "0.12", default-features = false }
 trie-root = { version = "0.12", default-features = false }
```
Comment (bkchr) [inline core/trie/Cargo.toml]:
```suggestion
trie-root = { version = "0.12.2", default-features = false }
```

File: core/trie/Cargo.toml [hunk 8]
```diff
@@ -12,11 +12,12 @@ name = "bench"
...
-memory-db = { version = "0.12", optional = true }
+memory-db = { version = "0.12", default-features = false }
```
Comment (bkchr) [inline core/trie/Cargo.toml]:
```suggestion
memory-db = { version = "0.12.2", default-features = false }
```

File: node-template/Cargo.toml [hunk 9]
```diff
@@ -19,7 +19,7 @@ exit-future = "0.1"
 parking_lot = "0.7.1"
 hex-literal = "0.1"
 parity-codec = "3.2"
-trie-root = "0.12.0"
+trie-root = "0.12"
```
Comment (bkchr) [inline node-template/Cargo.toml]:
```suggestion
trie-root = "0.12.2"
```

**Comment by bkchr** [comment]
Could you please call `use_trie` in a test and check the result? Please make sure the call is using wasm.

**Comment by cheme** [comment]
Yes, that was pretty useless without (except for triggering compilation error). I wonder if we want to use a new crate for wasm tests (put this one in sr-api-macros , the intent of the crate seems a bit off, but I would not really want to add the dependency for this test in trie).

**Comment by bkchr** [comment]
The dependency would only be `test-client` as `dev-dependency` to trie, but I'm also okay with the current solution. At least it is called :)

**Comment by bkchr** [comment]
Needs resolving and than we can merge :)

**Pull request merged**